### PR TITLE
feat(open_struct): query-layer wiring for OPEN_STRUCT columns (PR 3/4)

### PR DESCRIPTION
**Pull request by tarun11Mavani**
## Summary

This is **PR 3/4** in the OPEN_STRUCT stack — the query-engine changes that make `DataType.OPEN_STRUCT` columns queryable. Without it, declaring OPEN_STRUCT in a schema and ingesting data works (PR 2), but all queries crash or fall to per-doc expression scan.

RFC: https://docs.google.com/document/d/14kPmjDTKbO8l0ql4rrN7I5Yki5pqMw6GeGmxxc9grsU/edit?tab=t.0

PR 1 (SPI + data model): #18368 — merged
PR 2a (SPI utilities, index registration): #18710 — merged
PR 2b (storage layer): #18643 — merged

### What this PR adds

**Projection path (`SELECT myStruct['key']`):**
- `ItemTransformFunction` — accepts `OpenStructDataSource` alongside `MapDataSource`. When `getDataSource(key)` returns null (absent key), creates a typed `OpenStructNullDataSource` via `forAbsentKey()` factory (correct DataType from child FieldSpec, correct numDocs, full-segment null bitmap). Overrides `getNullBitmap()` to return the per-key DataSource's null bitmap directly — the base class ORs argument bitmaps, which doesn't work here because nullness depends on per-key presence, not parent column/literal nullness.
- `ProjectionBlock` — same pattern for path-based `[column, key]` value set access.

**Filter path (`WHERE myStruct['key'] = 'val'`):**
- `MapFilterOperator` — refactored from 2-way dispatch to 3-way:
  1. **PER_KEY_INDEX** (new) — for `OpenStructDataSource` columns. Uses `isMaterialized(key)` / `isFullyMaterialized()` for sub-dispatch:
     - Materialized key → rewrite predicate LHS, build `PredicateEvaluator`, delegate to `FilterOperatorUtils.getLeafFilterOperator` (picks inverted/range/sorted/scan automatically)
     - IS_NULL/IS_NOT_NULL → read `NullValueVectorReader` bitmap directly → `BitmapBasedFilterOperator`
     - Absent key (fully materialized) → IS_NULL → `MatchAllFilterOperator`, else → `EmptyFilterOperator`
     - Sparse key (not fully materialized) → fall through to expression scan
  2. **JSON_MATCH** (existing, unchanged) — MAP columns with JSON index
  3. **EXPRESSION_FILTER** (existing, unchanged) — fallback scan
- Single `_delegate` field replaces the old dual-field pattern. EXPLAIN output reports `delegateTo:per_key_index|json_match|expression_filter`.

**Aggregation path (`SELECT COUNT(DISTINCT myStruct['key'])`):**
- `AggregationPlanNode` — `resolveDataSource()` + `tryResolveKeyedDataSource()` resolve `item(col, 'key')` FUNCTION expressions to per-key DataSources. Enables dictionary/metadata-based non-scan aggregation (COUNT DISTINCT, MIN, MAX) on dense OPEN_STRUCT keys. `hasNullValues()` resolves per-key null vectors instead of pessimistically returning true for FUNCTION expressions.

**Absent key handling:**
- `ImmutableOpenStructDataSource.getDataSource(key)` — returns `null` for unmaterialized keys instead of falling back to the shared `$__sparse__` JSON column. The sparse column is an internal implementation detail used only by `getMapValue()`.
- `OpenStructNullDataSource` (new) — typed all-null DataSource for absent OPEN_STRUCT keys. Takes the correct `FieldSpec` (from `ComplexFieldSpec.getChildFieldSpec(key)`, falling back to STRING) and segment `numDocs`. Provides a typed forward index returning type-correct default null values, and a `NullValueVectorReader` with a full-segment bitmap.
- `PresenceBasedNullValueVector` (new) — `NullValueVectorReader` backed by `MutableKeyColumn.getPresenceBitmap()` complement. Exposes per-key null information on consuming segments so IS_NULL/IS_NOT_NULL work correctly. Clones the presence bitmap before full iteration to avoid concurrent mutation races with the ingestion thread.
- `MutableOpenStructDataSource.getDataSource(key)` — adds `PresenceBasedNullValueVector` to the per-key DataSource's index map.

**Segment loading:**
- `ImmutableSegmentImpl` — warn log on schema mismatch for OPEN_STRUCT columns.
- `MutableSegmentImpl` — `Preconditions.checkState` guard for OPEN_STRUCT columns with disabled index.
- `ImmutableOpenStructDataSource` — `getMapValue()` reconstructs the full map (dense + sparse) per document for seal path / minion tasks.

### Supported predicate types on per-key index path

EQ, NOT_EQ, IN, NOT_IN, RANGE, IS_NULL, IS_NOT_NULL. Other predicate types fall through to expression scan.

Known limitation: sparse OPEN_STRUCT keys read as NULL on the query path; sparse-column wiring will be added in next PR.

## Test plan

- [x] `MapFilterOperatorOpenStructTest` — 5 tests: EQ on materialized key, EQ on absent key (fully materialized) → Empty, IS_NULL on absent key → MatchAll, EQ on sparse key → expression fallback, IS_NOT_NULL with null bitmap → correct doc count
- [x] `AggregationPlanNodeResolveDataSourceTest` — 10 tests: identifier, literal, item() with MapDataSource, item() with materialized/unmaterialized OpenStructDataSource, non-item function, wrong arg count/types, plain DataSource
- [x] `OpenStructNullDataSourceTest` — 6 tests: all stored types (INT/LONG/FLOAT/DOUBLE/STRING/BYTES/BIG_DECIMAL), null vector covers all docs, zero-doc edge case, metadata correctness, no secondary indexes
- [x] `PresenceBasedNullValueVectorTest` — 5 tests: inverse of presence, complement computation, empty presence (all null), full presence (none null), zero docs
- [x] `ImmutableOpenStructDataSourceTest` — 14 tests: per-key DataSource, absent key returns null, isMaterialized, isFullyMaterialized, getMapValue dense+sparse, metadata
- [x] All existing pinot-core + pinot-segment-local tests pass
- [x] spotless/checkstyle/license clean

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java [hunk 1]
```diff
@@ -58,9 +60,17 @@ public void init(List<TransformFunction> arguments, Map<String, ColumnContext> c
...
-    Preconditions.checkState(dataSource instanceof MapDataSource, "Column: %s must be a MAP column", column);
-    MapDataSource mapDataSource = (MapDataSource) dataSource;
-    DataSource valueDataSource = mapDataSource.getDataSource(key);
+    Preconditions.checkState(dataSource instanceof MapDataSource || dataSource instanceof OpenStructDataSource,
+        "Column: %s must be a MAP or OPEN_STRUCT column", column);
+    DataSource valueDataSource;
+    if (dataSource instanceof MapDataSource) {
+      valueDataSource = ((MapDataSource) dataSource).getDataSource(key);
+    } else {
+      valueDataSource = ((OpenStructDataSource) dataSource).getDataSource(key);
```
Comment (xiangfu0) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java]: Sparse OPEN_STRUCT keys do not have a per-key DataSource here. `ImmutableOpenStructDataSource#getDataSource()` returns the shared `$__sparse__` JSON column for every unmaterialized key, so the expression fallback compares or projects the entire JSON blob instead of the requested value. For example, `metrics['region'] = 'us'` compares `{"region":"us",...}` with `us` and silently returns no rows. Only use the direct DataSource path for `isMaterialized(key)` and add a key-aware sparse extractor or JSON delegate.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java [hunk 2]
```diff
@@ -58,9 +60,17 @@ public void init(List<TransformFunction> arguments, Map<String, ColumnContext> c
     _keyPath = new String[]{column, key};
 
     DataSource dataSource = columnContextMap.get(column).getDataSource();
-    Preconditions.checkState(dataSource instanceof MapDataSource, "Column: %s must be a MAP column", column);
-    MapDataSource mapDataSource = (MapDataSource) dataSource;
-    DataSource valueDataSource = mapDataSource.getDataSource(key);
+    Preconditions.checkState(dataSource instanceof MapDataSource || dataSource instanceof OpenStructDataSource,
```
Comment (xiangfu0) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java:65]: The new OPEN_STRUCT path inherits `BaseTransformFunction#getNullBitmap()`, which only combines the null bitmaps of the parent `metrics` argument and key literal. The parent OPEN_STRUCT DataSource has no null vector, so the selected dense child's null bitmap is lost. Rows where the key is absent are emitted as default values and treated as non-null, causing wrong projections, composed expressions, and aggregates such as `COUNT(metrics['views'])` with null handling enabled. Override `getNullBitmap()` to return the selected path's BlockValSet null bitmap.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java [hunk 3]
```diff
@@ -58,9 +60,17 @@ public void init(List<TransformFunction> arguments, Map<String, ColumnContext> c
...
+        "Column: %s must be a MAP or OPEN_STRUCT column", column);
+    DataSource valueDataSource;
+    if (dataSource instanceof MapDataSource) {
+      valueDataSource = ((MapDataSource) dataSource).getDataSource(key);
+    } else {
+      valueDataSource = ((OpenStructDataSource) dataSource).getDataSource(key);
+    }
+    if (valueDataSource == null) {
+      valueDataSource = new NullDataSource(key);
```
Comment (xiangfu0) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java]: `NullDataSource` is not a type-correct all-null representation for an absent OPEN_STRUCT key. It is hard-coded as INT, has no null vector, and reports zero documents, so a schema-declared STRING or DOUBLE key missing from one segment becomes INT/0 there while other segments use its actual type. This can corrupt `COUNT`/`MIN`/`MAX` and produce incompatible per-segment result schemas. Construct a typed all-null DataSource from the OPEN_STRUCT child FieldSpec with a null bitmap covering the segment.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java [hunk 4]
```diff
@@ -67,104 +79,161 @@ public MapFilterOperator(IndexSegment indexSegment, Predicate predicate, QueryCo
...
+    return null;
+  }
+
+  @Nullable
+  private BaseFilterOperator buildPerKeyFilterOperator(DataSource keyDs, QueryContext queryContext, int numDocs) {
     switch (_predicate.getType()) {
+      case IS_NULL:
+      case IS_NOT_NULL: {
+        NullValueVectorReader nullReader = keyDs.getNullValueVector();
+        if (nullReader == null) {
```
Comment (xiangfu0) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java:194]: The assumption that a missing null vector means every value is non-null is false for consuming OPEN_STRUCT segments. `MutableKeyColumn` tracks per-key presence, but `MutableOpenStructIndex#getIndexes` exposes only forward, dictionary, and inverted indexes, so this null reader is absent even when the key exists in only some documents. `IS_NULL` consequently returns empty and `IS_NOT_NULL` matches every document. Expose a presence-backed `NullValueVectorReader` from mutable child DataSources before enabling this fast path.

**Review by xiangfu0 — COMMENTED** [review]
Found four high-severity correctness issues; see the inline comments.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java (hunk 1 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java]: `ImmutableOpenStructDataSource.getDataSource(key)` now returns null for unmaterialized keys instead of falling back to the sparse JSON column. Callers handle the null explicitly.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java (hunk 2 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java:65]: `ItemTransformFunction` now overrides `getNullBitmap()` to return the per-key DataSource's null bitmap directly.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java (hunk 3 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java]: Introduced `OpenStructNullDataSource` which takes the correct FieldSpec and numDocs, provides a typed forward index and a full-segment null bitmap. `NullDataSource` is left untouched for MAP columns.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java (hunk 4 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java:194]: `MutableOpenStructDataSource.getDataSource(key)` now adds a `PresenceBasedNullValueVector` (backed by `MutableKeyColumn.getPresenceBitmap()` complement) to the index map. The bitmap is cloned before iteration to avoid races with the ingestion thread.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java [hunk 5]
```diff
@@ -87,6 +103,11 @@ public Dictionary getDictionary() {
     return _dictionary;
   }
 
+  @Override
+  public RoaringBitmap getNullBitmap(ValueBlock valueBlock) {
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java:120]: This override now applies to MAP `item()` too, not just OPEN_STRUCT. On master `ItemTransformFunction` inherits `BaseTransformFunction.getNullBitmap`, which ORs the argument bitmaps (effectively never-null for a map key). Returning the per-key val-set bitmap silently changes `IS NULL` / null-aware results for existing `map['key']` queries. Was that MAP behavior change intended, and is it covered by a MAP-specific test?

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java [hunk 6]
```diff
@@ -204,6 +221,50 @@ private boolean isFitForNonScanBasedPlan() {
...
+    if (args.size() != 2
+        || args.get(0).getType() != ExpressionContext.Type.IDENTIFIER
+        || args.get(1).getType() != ExpressionContext.Type.LITERAL) {
+      return null;
+    }
+    String columnName = args.get(0).getIdentifier();
+    String key = args.get(1).getLiteral().getStringValue();
+    DataSource columnDs = segment.getDataSource(columnName, schema);
+    if (columnDs instanceof MapDataSource) {
+      return ((MapDataSource) columnDs).getDataSource(key);
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java]: This now makes `item(mapCol, 'k')` eligible for the non-scan aggregation path (previously any FUNCTION arg forced a scan). Two concerns for a key present in only some docs: (1) `BaseMapDataSource.getDataSource` returns a `NullDataSource` (INT, numDocs=0) for an absent key rather than null, so a non-scan COUNT/MIN/MAX could run over 0 docs; (2) the per-key MAP source may carry no null vector, so `hasNullValues` returns false and non-scan agg would ignore absent-doc nulls. Can you confirm the per-key MAP source models absent-doc nulls / full segment numDocs here?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/PresenceBasedNullValueVector.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  private final int _numDocs;
+
+  public PresenceBasedNullValueVector(ImmutableRoaringBitmap presenceBitmap, int numDocs) {
+    _presenceBitmap = presenceBitmap;
+    _numDocs = numDocs;
+  }
+
+  @Override
+  public boolean isNull(int docId) {
+    return !_presenceBitmap.contains(docId);
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/PresenceBasedNullValueVector.java:54]: `getNullBitmap()` below clones the presence bitmap before iterating to avoid racing the ingestion thread, but `isNull()` reads the live mutable bitmap directly via `contains()`. On a consuming segment, isn't this the same concurrent-mutation hazard the clone was added to guard against? Should isNull go through the same protection (or is contains() safe against a concurrent `add`)?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/ImmutableOpenStructDataSource.java [hunk 8]
```diff
@@ -84,8 +84,7 @@ public ComplexFieldSpec getFieldSpec() {
   @Override
   @Nullable
   public DataSource getDataSource(String key) {
-    DataSource ds = _perKeyDataSources.get(key);
-    return ds != null ? ds : _sparseDataSource;
+    return _perKeyDataSources.get(key);
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/ImmutableOpenStructDataSource.java:91]: This flips `getDataSource(key)` from returning the `$__sparse__` fallback to returning `null` for unmaterialized keys — a semantic change to a method other call sites may have assumed non-null. Have all existing callers been audited to handle null (vs. the old sparse fallback)?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java [hunk 9]
```diff
@@ -62,10 +63,13 @@ public ComplexFieldSpec getFieldSpec() {
   @Override
   @Nullable
   public DataSource getDataSource(String key) {
-    Map<IndexType, IndexReader> indexes = _index.getIndexes(key);
-    if (indexes == null || indexes.isEmpty()) {
+    MutableKeyColumn col = _index.getKeyColumn(key);
+    if (col == null) {
       return null;
     }
+    Map<IndexType, IndexReader> indexes = new HashMap<>(_index.getIndexes(key));
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java]: This allocates a fresh `HashMap` copy plus a new `PresenceBasedNullValueVector` on every `getDataSource(key)` call, and `getNullBitmap()` then rebuilds a full `[0,numDocs)` bitmap + clone per invocation. If this is hit per-predicate on the query path, is it worth caching the per-key DataSource, or is getDataSource already called once per query?

**Review by raghavyadav01 — COMMENTED** [review]
A few questions on the query-layer wiring, mostly around the MAP path this overlaps with (null-bitmap semantics, non-scan aggregation eligibility) plus one thread-safety and one contract question. Details inline.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java [hunk 10]
```diff
@@ -67,104 +79,161 @@ public MapFilterOperator(IndexSegment indexSegment, Predicate predicate, QueryCo
...
+      return buildPerKeyFilterOperator(keyDs, queryContext, numDocs);
+    }
+
+    // Key not materialized
+    if (osDs.isFullyMaterialized()) {
+      // Fully materialized but key absent — definitive answer
+      if (_predicate.getType() == Predicate.Type.IS_NULL) {
+        return new MatchAllFilterOperator(numDocs);
+      }
+      return EmptyFilterOperator.getInstance();
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java]: For a fully-materialized-but-absent key this hardcodes `IS_NULL → MatchAll` and everything else → `Empty`. With null handling disabled, `NOT_EQ` / `NOT_IN` against an all-default column should match *all* docs, so `Empty` looks wrong for those. Rather than special-casing here, could this delegate to `buildPerKeyFilterOperator(OpenStructNullDataSource.forAbsentKey(osDs, _keyName), ...)`? That's the typed all-null source (with a full null vector) this PR already builds for the projection path, and it'd keep EQ/NOT_EQ/IS_NULL semantics consistent across filter and projection.

**Review by raghavyadav01 — COMMENTED** [review]
One more on the filter path — absent-key handling vs. NOT_EQ/NOT_IN under disabled null handling. Inline.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java [hunk 11]
```diff
@@ -62,10 +63,13 @@ public ComplexFieldSpec getFieldSpec() {
...
+    indexes.put(StandardIndexes.nullValueVector(),
+        new PresenceBasedNullValueVector(col.getPresenceBitmap(), _numDocs));
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java]: The mutable path synthesizes a `PresenceBasedNullValueVector` here so per-key null info is exposed. On the sealed side, `ImmutableSegmentImpl` builds each materialized key as a plain `ImmutableDataSource`, so `getNullValueVector()` returns only what was written at seal time — there's no equivalent synthesis. For a **dense key that's materialized but present in only some docs**, does the segment build write a null vector for the absent docs? If not, on immutable segments `buildPerKeyFilterOperator`'s `nullReader == null` branch would treat those docs as non-null (IS_NULL → Empty, IS_NOT_NULL → MatchAll), and `hasNullValues` would return false and let a non-scan aggregation ignore them — both wrong. Worth a test with a partially-present dense key that survives seal, run on the immutable segment.

**Review by raghavyadav01 — COMMENTED** [review]
Correctness question on mutable-vs-immutable null-vector parity for partially-present dense keys. Inline.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java (hunk 5 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/ItemTransformFunction.java:120]: You're right that the base behavior is correct for MAP: there's no per-key null bitmap, so ORing the argument bitmaps yields the whole-map bitmap, a conservative over-estimate. 
 
Removing the override entirely doesn't work either: for OPEN_STRUCT the parent DataSource is built with an empty index container (`ImmutableOpenStructDataSource:84`, `MutableOpenStructDataSource:51`), so the base impl would report "no nulls" there too. And `TransformBlockValSet.getNullBitmap()` delegates straight to the transform function (`:60`), so this is the only place per-key nulls can come from.
 
Scoped by column type instead: OPEN_STRUCT reads the per-key bitmap, MAP delegates to `super`. Added MAP tests including the absent-key case — both fail against the previous unconditional override.

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java (hunk 6 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java]: Checked both concerns. Neither can bite today, but I've added the guard.
 
(1) **COUNT** never reaches the per-key DataSource — `isFitForNonScanBasedPlan` skips COUNT before resolving it, and the operator counts from segment metadata. So `NullDataSource`'s `numDocs=0` is never read.
 
(2) **MIN/MAX** is the real hole, and it's min/max rather than the null vector: `NullDataSourceMetadata` returns `Integer.MIN_VALUE` for both (`NullDataSource:181-189`), so an absent MAP key would pass the metadata check and `MIN(mapCol['absentKey'])` would return `-2147483648` instead of NULL.
 
That path is unreachable on master, though. `BaseMapDataSource` only returns a `NullDataSource` when `getMapIndexReader().getIndexes(key)` is null, and the only `MapIndexReader` implementation returns a non-null map unconditionally. Absent keys land on `ImmutableDataSource` instead, whose min/max are null, so they fail the metadata check and fall back to a scan.
 
Added the guard regardless — `return keyDs instanceof NullDataSource ? null : keyDs`, mirroring the OPEN_STRUCT branch. `ItemTransformFunction` already null-checks the same call, so the nullable contract is clearly intended; this just makes the aggregation path honour it too.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/PresenceBasedNullValueVector.java (hunk 7 above)
Comment (tarun11Mavani) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/PresenceBasedNullValueVector.java:54]: thanks for flagging this. Replaced  ImmutableRoaringBitmap with ThreadSafeMutableRoaringBitmap.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/ImmutableOpenStructDataSource.java (hunk 8 above)
Comment (tarun11Mavani) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/ImmutableOpenStructDataSource.java:91]: Yes. This is an intentional change. 
The reason for the flip: `_sparseDataSource` is one JSON blob shared by every sparse key, not the
key's values, and it has the wrong type. Returning it meant a caller couldn't distinguish "key
absent" from "key lives in sparse", so it couldn't choose a fallback.

I have updated the SPI method declaration and the java doc for more clarity.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java (hunk 10 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/MapFilterOperator.java]: You're right — with null handling off an absent key reads as its type default (`"null"` for STRING),
so `NOT_EQ` / `NOT_IN` should match every doc. Fixed, routed through `forAbsentKey` as you suggested.
 
One deviation: delegating to `buildPerKeyFilterOperator` lands on a scan of all N docs for a constant
answer and loses `canOptimizeCount()`, so I apply the evaluator once and fold to `MatchAll`/`Empty`.
Same result, O(1).
 
Three other behavior changes fall out of this, all consistent with "absent key reads as its default":
`RANGE` is evaluated rather than always `Empty` (lexicographic for undeclared keys, which fall back
to STRING), `REGEXP_LIKE` falls through to the expression path instead of `Empty`, and a type
mismatch like `item(m,'absentLongKey') = 'abc'` now throws instead of returning 0 rows — same as a
materialized LONG key.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java (hunk 11 above)
Comment (tarun11Mavani) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java]: Good catch on the asymmetry. the two sides do get null info by different mechanisms — but the
sealed side is covered. `OpenStructColumnSplitter:326-336` runs a `NullValueVectorCreator` for every
materialized child column at seal time, so dense keys get a real on-disk null vector and the
`nullReader == null` branch isn't reached.
 
No existing test has a partially-present dense key. I plan to add a dedicated pr with integration tests. I will cover this case as well.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java [hunk 12]
```diff
@@ -62,13 +69,19 @@ public ComplexFieldSpec getFieldSpec() {
...
+    // Live lookup, outside the memo: a key not yet observed may still be created by the ingestion thread.
+    MutableKeyColumn col = _index.getKeyColumn(key);
+    if (col == null) {
       return null;
     }
-    ColumnMetadata metadata = _index.getColumnMetadata(key);
-    return new ImmutableDataSource(metadata,
-        new ColumnIndexContainer.FromMap.Builder().withAll(indexes).build());
+    return _perKeyDataSourceCache.computeIfAbsent(key, k -> {
+      Map<IndexType, IndexReader> indexes = new HashMap<>(_index.getIndexes(k));
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java:80]: This hands the engine `MutableKeyColumn`'s raw forward index, which only has slots written at present docIds (1000-row chunks). A full `[0, numDocs)` projection/scan reads via `_readers.get(docId / 1000)` with no presence check — so wouldn't an absent doc past the last present chunk hit `_readers.get(bufferId)` out-of-bounds, and an in-range absent doc return dictId `0` (the first inserted value)? Should the per-key read be presence-aware (default for absent docs) here, and do we have a consuming-segment test that projects/scans a key absent from the tail rows?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java [hunk 13]
```diff
@@ -59,15 +63,37 @@ public BlockValSet getBlockValueSet(ExpressionContext expression) {
...
+    if (columnDataSource instanceof MapDataSource) {
+      keyDataSource = ((MapDataSource) columnDataSource).getDataSource(paths[1]);
+      if (keyDataSource == null) {
+        keyDataSource = new NullDataSource(paths[1]);
+      }
+    } else if (columnDataSource instanceof OpenStructDataSource) {
+      OpenStructDataSource osDs = (OpenStructDataSource) columnDataSource;
+      keyDataSource = osDs.getDataSource(paths[1]);
+      if (keyDataSource == null) {
+        keyDataSource = OpenStructNullDataSource.forAbsentKey(osDs, paths[1]);
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java:98]: For a sparse (not-yet-materialized) key, `getDataSource` returns null even though the value may still live in the sparse blob, so this projects all-NULL for every row. Should sparse keys fall back to the sparse/JSON read path here, or is materialized-only the intended scope for this PR?

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java [hunk 14]
```diff
@@ -204,6 +222,55 @@ private boolean isFitForNonScanBasedPlan() {
...
+      DataSource keyDs = ((MapDataSource) columnDs).getDataSource(key);
+      // An absent MAP key yields a NullDataSource whose metadata reports non-null min/max (the INT
+      // default) and which carries no null vector, so it would wrongly satisfy the metadata-based
+      // non-scan check and report hasNullValues=false. Fall back to a scan, mirroring the
+      // OPEN_STRUCT guard below.
+      return keyDs instanceof NullDataSource ? null : keyDs;
+    }
+    if (columnDs instanceof OpenStructDataSource) {
+      OpenStructDataSource osDs = (OpenStructDataSource) columnDs;
+      return osDs.isMaterialized(key) ? osDs.getDataSource(key) : null;
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java]: For a key present in only some docs, this routes MIN/MAX/DISTINCTCOUNT to the non-scan path, which reads the dictionary — and the mutable dictionary holds present values only. With null handling off, doesn't that diverge from the scan path (and from sealed segments, which fold the default null in at build time)? Might be worth a consuming-vs-sealed test to pin the expected value.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java (hunk 12 above)
Comment (tarun11Mavani) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/openstruct/MutableOpenStructDataSource.java:80]: Confirmed both failure modes: docIds past the last written chunk hit `_readers.get(bufferId)` OOB, and in-range holes deserialized as dictId 0 (first inserted value). Fixed by making dictId 0 *be* the answer: the dictionary now reserves it for the key's default null value at construction, so zero-initialized chunks read absent docs as the default (same value the sealed build folds in), and a last-written-docId watermark guards reads past the allocated tail. 
Added `OpenStructConsumingSealedParityTest` covering a key absent from middle + tail rows, consuming vs sealed.

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java (hunk 14 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java]: Confirmed. With the reserved default in the mutable dictionary, a partially-present key's dictionary now has identical contents to the sealed one (default + observed values), so the dictionary-based path is correct there. 

The one divergent case — key present in every doc with the default never observed (phantom entry)  forces the scan path via a new `OpenStructDataSource.isKeyDictionaryExact`  
The parity test pins MIN/MAX/DISTINCTCOUNT consuming vs sealed with null handling off.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java (hunk 13 above)
Comment (tarun11Mavani) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/ProjectionBlock.java:98]: this PR scope if Materialized-only. 
I am planning to add support for sparse key query including json_index for those keys in upcoming PR.

File: pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java [hunk 15]
```diff
@@ -187,6 +191,15 @@ private boolean hasNullValues(AggregationFunction[] aggregationFunctions) {
             }
             break;
           case FUNCTION:
+            DataSource resolvedDs = resolveDataSource(argument);
+            if (resolvedDs == null) {
+              return true;
+            }
+            NullValueVectorReader resolvedNullVector = resolvedDs.getNullValueVector();
+            if (resolvedNullVector != null && !resolvedNullVector.getNullBitmap().isEmpty()) {
```
Comment (raghavyadav01) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java:199]: This narrows what used to be a blanket "any FUNCTION arg might be null" assumption. Now `item()` on a materialized key whose per-key null vector is empty (or null) falls through to `break`, so the aggregation is treated as null-free and takes the non-scan path. That's only correct if the per-key null vector is authoritative for absent docs — which is exactly the mutable-vs-sealed parity question raised earlier in the review.

Not blocking, but could we add a null-handling-**on** test that pins this: `COUNT`/`MIN`/`MAX` over a partially-present key, run on both a consuming and a sealed segment? The existing parity test only covers null handling off.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/RealtimeInvertedIndex.java [hunk 16]
```diff
@@ -57,6 +57,19 @@ public void add(int dictId, int docId) {
     }
   }
 
+  /// Pre-creates an empty bitmap for the next dictionary id. Used by callers that reserve a
+  /// dictionary id upfront (e.g. OPEN_STRUCT key columns reserve dictId 0 for the default null
+  /// value) so that subsequent contiguous [#add] calls stay aligned with dictionary ids.
+  public void reserveNextDictId() {
```
Comment (raghavyadav01) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/invertedindex/RealtimeInvertedIndex.java:63]: `reserveNextDictId()` appends unconditionally, unlike `add(dictId, docId)` which guards on `_bitmaps.size() == dictId`. It's safe today because `MutableKeyColumn` calls it exactly once on an empty index, but the method is public and generically named — a second call, or an interleave with `add()`, would silently misalign every dictId→docId bitmap after it. Worth a cheap guard, e.g. `Preconditions.checkState(_bitmaps.isEmpty())`, to make the "only on an empty index" contract enforced rather than assumed.

**Review by raghavyadav01 — APPROVED** [review]
LGTM. The query-layer wiring is solid and CI is green. Two non-blocking follow-ups inline: (1) a null-handling-on parity test for aggregations over a partially-present key (consuming + sealed), and (2) a defensive guard on `reserveNextDictId()`. Neither blocks this PR — happy to see them in the next one.

**Comment by xiangfu0** [comment]
Documentation follow-up: https://github.com/pinot-contrib/pinot-docs/pull/961
